Allow for "deep-copy" of a dashboard

dpsutton opened the pull request:
the existing copy dashboard url (dashboard/:id/copy) now takes two new arguments:

- destination-collection: a collection id to put the copies and dash
- copy-style: "deep" or "shallow" (omitting assumes prevoius behavior of "shallow"

Todo:
- [x] multi-series
- [x] text cards
- [x] validation
  - [x] destination collection exists
  - [x] user can read destination collection
- [x] reuse metadata when copying card
- [x] handle when source and destination collections are the same (append " -- Duplicate" to title)
- [x] don't copy cards if you lack ~curate~ read permissions
  - [x] don't copy series cards you _do_ have permission for if the base card lacks permissions
- [x] handle dashboard filters 
- [x] delay `card-create` event until transaction creating all cards and dashboard completes

dpsutton committed:
Allow for "deep-copy" of a dashboard

the existing copy dashboard url (dashboard/:id/copy) now takes two new
arguments:

- destination-collection: a collection id to put the copies and dash
- copy-style: "deep" or "shallow" (omitting assumes prevoius behavior of "shallow"

dpsutton committed:
Handle text cards

dpsutton committed:
lint fix

dpsutton committed:
move text back inside viz settings

dpsutton committed:
Ensure destination collection exists

dpsutton committed:
Add logging and use passed in metadata

dpsutton committed:
Handle permissions and same collection deep copy

- Filter out cards which user does not have curate collection
permissions
because of how cards work, `(mi/can-read? card)` is the same as curate
permission on the collection. So each card is removed which you lack
curate on. In the case of series, we remove each series card if you lack
write, and do not copy the series at all if the base card lacks
permissions, since there's no where to stack them.

Data looks like:
{:card <a card>
 :series [<a card> <a card>]}

This arrangement is where there are multiple line graphs stacked on top
of each other. If the base card can't be copied, there's no where for
the stacked cards to be displayed so none are copied, even if you have
permissions to the cards stacked on top.

- Same collection deep copy
If you copy to the same collection, each card's name is updated to
include " -- Duplicate".

dpsutton committed:
Handle parameter_mappings in dashboards

Scenario:
dashboard filters live top-level on the dashboard

```clojure
 :parameters [{:name "Text",
               :slug "text",
               :id "72d27de6",
               :type :string/=,
               :sectionId "string"}]
```

And then each dashboard card will connect the filter to the card on how
to apply the filter (choose "Gizmo" in the dropdown, the card should
include where category = 'Gizmo'). That data looks like

```clojure
:parameter_mappings [{:parameter_id "72d27de6",
                      :card_id 809,
                      :target [:dimension [:field 63 nil]]}]
```

And we need to ensure we update the card_id there is updated.

But its not clear to me why we store it there. Observe the following
query:

```sql
select id, card_id, parameter_mappings from report_dashboardcard where dashboard_id = 329;
 id  | card_id |                parameter_mappings
-----+---------+--------------------------------------------------------
 430 |     809 | [{"parameter_id":"72d27de6","card_id":809,
     |         |   "target":["dimension",["field",63,null]]}]
 437 |     816 | [{"parameter_id":"72d27de6","card_id":816,
     |         |   "target":["dimension",["template-tag","CATEGORY"]]}]
 ```

Each dashboard card includes the card_id in the parameter mappings but
also in the schema. (related to series maybe?) or perhaps just
duplication for no reason.

gusaiani commented:
@dpsutton, frontend is now able to duplicate a dashboard deep and shallow.

Onwards to details.

Today we pass "collection_id" to the `copy` endpoint, and using it the dashboard goes to the corresponding collection.
What do you think of simply using "collection_id" as the place to save the duplicated questions etc as well?

(or maybe I didn't quite understand the difference between "collection_id" and "destination-collection")

gusaiani commented:
> (or maybe I didn't quite understand the difference between "collection_id" and "destination-collection")

Also, as I typed the above, shall we continue snake_casing these attribute names?

gusaiani commented:
@dpsutton, if we want to make FE code super simple in telling the BE if the copy is shallow or deep, we could rename the "copy-style" attribute to something like "is_shallow_copy", which gets a boolean, with the default for "false".

This would work seamlessly with how checkboxes are generated in our codebase.

What do you think, is this attractive to you?

Loom of the same where our code handles this in the FE: https://www.loom.com/share/735009d6366a4497a392f52753b5507b

dpsutton commented:
works for me. I'll update the BE to work in this manner.

dpsutton committed:
Gus's changes

- just use `collection_id` from the body rather than the new bespoke thing
- use `is_deep_copy` for indicating if deep or shallow copy

`is_deep_copy` defaults to nil/false in the BE. The FE should default to
sending true. This makes the new behavior the default in the application
but preserves the existing behavior for any users of the api.

dpsutton committed:
remove redundant perm check on collection_id

was put in when destination collection was a separate thing. and we
already do the more correct
`collection/check-write-perms-for-collection` which handles the root
collection as well.

dpsutton committed:
Don't `:refer` Collection any longer

dpsutton committed:
Update `update-cards-for-copy` to take a boolean for copy -style

also assert that if we are deep copying we have to have a mapping from
old to new card ids. Should work even for empty dashboards (it only
asserts it is a map, not that it has any bindings in it). This is nil in
the shallow copy pathway.

gusaiani committed:
Frontend for deep copying dashboards (#25949)

dpsutton committed:
Include uncopied cards on deep-copy response

- cards-to-copy tracks cards copied and uncopied now
- dashboard response includes a key `:uncopied` with is a vector of
  cards not copied
- Makes `metabase.models.dispatch/Model` protocol extendable via
  metadata. Allows for nicer testing where I can mock out
  `mi/can-write?` instead of actually nesting into collections,
  permissions, etc.

gusaiani committed:
Update duplicate modal title if dashboard copy is set to deep or shallow (#26021)

gusaiani committed:
Deep-copying dashboards — customize checkbox label (#26027)

gusaiani committed:
Update modal title if deep/shallow from collection page (#26035)

gusaiani committed:
Extract getTitle functions from component (#26060)

qnkhuat commented on src/metabase/api/dashboard.clj:
```diff
@@ -215,14 +216,97 @@
-   collection_id       (s/maybe su/IntGreaterThanZero)
-   collection_position (s/maybe su/IntGreaterThanZero)}
+  [from-dashboard-id :as {{:keys [name description collection_id collection_position
+                                  is_deep_copy], :as _dashboard} :body}]
```
```suggestion
                                  is_deep_copy], :as _dashboard, :or {is_deep_copy false}} :body}]
```

Might be a good idea to have a default value here. It's better for readability IMO.

dpsutton replied:
i find nil destructuring just as readable. no strong opinion but this feels redundant to me. Of course the default value is nil.

qnkhuat commented on src/metabase/api/card.clj:
```diff
@@ -208,25 +208,38 @@
             valid-metadata?))
-      (a/to-chan! [metadata])
+      (do
+        (log/debug (trs "Reusing provided metadata"))
```
do we wish to keep these `log/debug`? Just double-checking because it uses i18n and I think we don't want too many translations for things we don't really need.

dpsutton replied:
I see no downside to having the string. And when you want this information it is quite nice to have.

qnkhuat commented on src/metabase/api/dashboard.clj:
```diff
@@ -215,14 +216,97 @@
+  id from the newly-copied cards. If there is no new id, it means user lacked curate permissions for the cards
+  collections and it is omitted. Dashboard-id is only needed for useful errors."
+  [dashboard-id ordered-cards deep? id->new-card]
+  (when (and deep? (nil? id->new-card))
```
Is it better if we move this check to the API?
So that this function can be changed to `(defn reconcile-cards-for-deep-copy [ordered-cards id->new-card])` 
where its sole job is to replace ids from old to new IDs?

dpsutton replied:
I don't think so. The `id->new-card` is a map from the old id to the copied cards. At this point, if we're doing a shallow copy, there is no work to do. Otherwise we need to update the old ids to be the new ids. So this check is just if we are doing a deep copy, it is an error to not have the copied cards to substitute in. So I think this check here makes sense.

qnkhuat commented on src/metabase/api/dashboard.clj:
```diff
@@ -215,14 +216,97 @@
+        {:keys [copy discard]} (cards-to-copy (:ordered_cards dashboard))]
+    (reduce (fn [m [id card]]
+              (assoc m id
+                     (if (:dataset card)
```
could you explain why we don't duplicate the model?

dpsutton replied:
Consider models to be tables. They are more permanent that just cards. They also might be persisted to the source database and we just copy a reference to them.

qnkhuat commented:
I found a bug, please check the view below for how to repoduce
https://user-images.githubusercontent.com/25661381/198216263-6a2cb985-b54b-4f3f-8d18-576e8511efbe.mov


The exception is thrown from this line : 

https://github.com/metabase/metabase/blob/deep-copy-dashboards/src/metabase/api/dashboard.clj#L228 

I did some debugging and I think I found out why, 
the reason we got this exception is the `card` when we call `(mi/can-read card)`   is not an Instance of `metabase.models.card/Card`. In fact it's a map with only `id` as the key.

And I think this is where we change from `CardInstace` to a `map`.

https://github.com/metabase/metabase/blob/deep-copy-dashboards/src/metabase/api/dashboard.clj#L103-L109

qnkhuat reviewed (commented):
Having some comments and also found a small bug.

I'm happy to pair in case any of my comments are unclear.

bshepherdson commented on src/metabase/api/dashboard.clj:
```diff
@@ -215,14 +216,97 @@
+                (-> dashboard-card
+                    (update :card_id new-id)
+                    (assoc :card (-> dashboard-card :card_id id->new-card))
+                    (m/update-existing :parameter_mappings
```
What about the case where the card's query depends on another card, with `{:dataset_query {:query {:source_table "card__1234"}}}`? Perhaps those are treated like models, and we don't want to duplicate them. I just wanted to make sure it was considered.

dpsutton replied:
I don't think that's a problem. If they can edit the collection where the nested query is then its good to go. The UI on the individual card will offer to duplicate it, etc. Good call out though.

dpsutton commented on src/metabase/api/dashboard.clj:
```diff
@@ -215,14 +216,97 @@
+                    (m/update-existing :parameter_mappings
+                                       (fn [pms]
+                                         (map (fn [pm]
+                                                (m/update-existing pm :card_id new-id))
```
good call. I'll change the `map` to `keep`

dpsutton committed:
keep parameter_mappings, not map

dpsutton committed:
Switch to `can-read?` not `can-write?`

original requestor intends to use a read-only template to be
copied. Thus no one can write to it (its a template, keep it clean) so
no one would be able to copy it, defeating the purpose.

Further, the ability to copy cards is based on `can-read?` so it wasn't
a meaningful or consistent protection:

```clojure
;; POST api/card/<id>/copy
;; based on `read-check`, not `write-check`
(api/defendpoint POST "/:id/copy"
  "Copy a `Card`, with the new name 'Copy of _name_'"
  [id]
  {id (s/maybe su/IntGreaterThanZero)}
  (let [orig-card (api/read-check Card id)
        new-name  (str (trs "Copy of ") (:name orig-card))
        new-card  (assoc orig-card :name new-name)]
    (create-card! new-card)))
```

dpsutton committed:
Delay firing `:card-created` event until after tx

repeatedly calling `create-card!` fires the card created
event. Principally what this does is create the first revision of the
new card. But this work cannot find the newly created card until the
`db/transaction` is complete. And this transaction does not complete
until the new dashboard is fully created with all cards.

So just keep a copy of the new cards and fire the event after the tx
completes

dpsutton committed:
Check `mi/instance-of? Card card` to just `mi/model`

tests use a fake model who's permissions are subbed to make it easy to
say which ones are `mi/can-read?`. But they are not Card models. So just
check if they are models at all (the test ones are
`::dispatches-on-dynamic` models who consult a dynamic set of readable
ids) instead of hitting collection id and checking user perms. Much
easier to fake in testing.

qnkhuat commented on test/metabase/api/dashboard_test.clj:
```diff
@@ -690,7 +693,335 @@
+                                               :collection_id (u/the-id source-coll)
+                                               :display :line
+                                               :visualization_settings
+                                               {:graph.dimensions ["CREATED_AT"]
```
nit: Correct me if I'm wrong but  `display` and `visualization_settings` are not used in the tests, so maybe we could delete them? 
It makes the tests more focused IMO.

dpsutton replied:
fair point. I set this up so that the dashboard is actually a well-formed dashboard. If you pause the tests halfway through running the cards have line graphs and some are stacked multi-series. certainly true there are no assertions on the settings at the moment. I'll think about that and see if i can/should assert on that.

qnkhuat commented on src/metabase/api/dashboard.clj:
```diff
@@ -234,16 +325,29 @@
+                        ;; Ok, now save the Dashboard
+                        (let [dash (db/insert! Dashboard dashboard-data)
+                              id->new-card (when is_deep_copy
+                                             (duplicate-cards existing-dashboard collection_id))]
```
nit: I think it'll be clearer if `duplicate-cards` return 2 keys `[:copied :uncopied]` rather than `[id1, id2,.. :uncopied]`

dpsutton replied:
done. good point.

qnkhuat reviewed (approved):
This LGTM, I have some small comments but it shouldn't block from merging. Up to you whether or not to resolve it.

gusaiani committed:
After deep-copying dashboard, warn in toaster if unduplicated questions (#26161)

dpsutton committed:
Clean up return value of `duplicate-cards`

Used to return a map of id->copied card which also included a key
`:uncopied`. Now is a map of two keys, :copied, and :uncopied, so that
it's homogenous and not just mixed together.
